### PR TITLE
add mobile custom error

### PR DESCRIPTION
**Pull request by aherzberg**
## Summary
https://department-of-veterans-affairs.github.io/va-mobile-app/docs/Engineering/BackEnd/Features/CustomErrorResponses/

https://adhoc.slack.com/archives/C0212B872MT/p1717780791608929

## Related issue(s)
https://github.com/department-of-veterans-affairs/va-mobile-app/issues/8810

## Testing done

- [ ] *New code is covered by unit tests*
- *Describe what the old behavior was prior to the change*
- *Describe the steps required to verify your changes are working as expected. Exclusively stating 'Specs run' is NOT acceptable as appropriate testing*
- *If this work is behind a flipper:*
  - *Tests need to be written for both the flipper on and flipper off scenarios. [Docs](https://depo-platform-documentation.scrollhelp.site/developer-docs/feature-toggles-guide#Featuretogglesguide-Backendexample).*
  - *What is the testing plan for rolling out the feature?*

## Screenshots
_Note: Optional_

## What areas of the site does it impact?
*(Describe what parts of the site are impacted and*if*code touched other areas)*

## Acceptance criteria

- [ ]  I fixed|updated|added unit tests and integration tests for each feature (if applicable).
- [ ]  No error nor warning in the console.
- [ ]  Events are being sent to the appropriate logging solution
- [ ]  Documentation has been updated (link to documentation)
- [ ]  No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs
- [ ]  Feature/bug has a monitor built into Datadog or Grafana (if applicable)
- [ ]  If app impacted requires authentication, did you login to a local build and verify all authenticated routes work as expected
- [ ]  I added a screenshot of the developed feature

## Requested Feedback

(OPTIONAL)_What should the reviewers know in addition to the above. Is there anything specific you wish the reviewer to assist with. Do you have any concerns with this PR, why?_

### CODE REVIEW

File: modules/mobile/app/controllers/mobile/v0/appointments_controller.rb [hunk 1]
```diff
@@ -9,6 +11,15 @@ class AppointmentsController < ApplicationController
       after_action :clear_appointments_cache, only: %i[cancel create]
 
       def index
+        if Settings.vsp_environment != 'production' && @current_user.email == 'vets.gov.user+141@gmail.com'
```
Comment (kpethtel) [inline modules/mobile/app/controllers/mobile/v0/appointments_controller.rb:14]: I think you should move this to a private method.

File: modules/mobile/lib/mobile/v0/exceptions/custom_errors.rb [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+          @refreshable = refreshable
+          super
+        end
+
+        def errors
+          [
+            {
+              title: @title,
+              body: @body,
+              status: 418,
```
Comment (kpethtel) [inline modules/mobile/lib/mobile/v0/exceptions/custom_errors.rb:25]: Status should be populated from the locales. I think something like `i18n_data[:status]` will work.

**Review by kpethtel — CHANGES_REQUESTED** [review]
Is this intended to be permanent or long lived within appointments?

I think we should also add a test within the appointments request specs confirming that it works for the test user and returns 418 with the expected data.

File: modules/mobile/lib/mobile/v0/exceptions/custom_errors.rb (hunk 2 above)
Comment (aherzberg) [reply modules/mobile/lib/mobile/v0/exceptions/custom_errors.rb]: the status won't ever change. I've moved it to a constant but don't want to give a user adding this error the opportunity to fat thumb the status code

**Comment by aherzberg** [comment]
> Is this intended to be permanent or long lived within appointments?
> 
> I think we should also add a test within the appointments request specs confirming that it works for the test user and returns 418 with the expected data.

I'm not sure how long QA will want this code in place to test. They may want it permanently as their go to user/endpoint to test this custom response.

File: modules/mobile/lib/mobile/v0/exceptions/custom_errors.rb (hunk 2 above)
Comment (kpethtel) [reply modules/mobile/lib/mobile/v0/exceptions/custom_errors.rb]: I think we may be talking about different things. I'm saying that by using the i18n method, you'll be using the error data you added to `modules/mobile/config/locales/exceptions.en.yml`. Which effectively acts as a constant. This is part of the established pattern that vets-api error classes use.

Comment (aherzberg) [reply modules/mobile/lib/mobile/v0/exceptions/custom_errors.rb]: oh got it. sure, added `i18n_data[:status]`

**Comment by kpethtel** [comment]
> > Is this intended to be permanent or long lived within appointments?
> > I think we should also add a test within the appointments request specs confirming that it works for the test user and returns 418 with the expected data.
> 
> I'm not sure how long QA will want this code in place to test. They may want it permanently as their go to user/endpoint to test this custom response.
> 
> For the request spec, I had a request spec to confirm it did work but didn't push it since I thought we were trying to move away from having every edge case tested in request specs. I'm happy to add it back in but not sure where to draw the line between the two.

@aherzberg There are a couple of reasons I think this should be tested. 
1) as a reviewer, it's not 100% clear to me that this will really return a 418 status. If you tested it and feel confident, I'm ok with not adding a permanent test. 
2) This isn't an edge case. It's the first line of the index method and completely changes behavior. It also can't be tested elsewhere. What I've been harping about is establishing appropriate testing boundaries, moving business logic into classes, and doing the nitty gritty tests within unit tests on those classes. For example, we have an appointments adapter spec for confirming that it handles all the conditional logic for transforming the appointment, so that doesn't need to be done in the request spec. But the request specs still need to test the high level stuff like: 
- when the upstream returns 200, we return 200 with the correct data
- when some upstreams succeed and others fail, we return 207 with some data and an error detail
Those things can't really be tested elsewhere because they involve controller behaviors like error handling. This is a big subject and it's not always clear what should be integration tested and what shouldn't. Maybe we can talk about it in our next monthly team meeting.

**Comment by aherzberg** [comment]
> > > Is this intended to be permanent or long lived within appointments?
> > > I think we should also add a test within the appointments request specs confirming that it works for the test user and returns 418 with the expected data.
> > 
> > 
> > I'm not sure how long QA will want this code in place to test. They may want it permanently as their go to user/endpoint to test this custom response.
> > For the request spec, I had a request spec to confirm it did work but didn't push it since I thought we were trying to move away from having every edge case tested in request specs. I'm happy to add it back in but not sure where to draw the line between the two.
> 
> @aherzberg There are a couple of reasons I think this should be tested.
> 
> 1. as a reviewer, it's not 100% clear to me that this will really return a 418 status. If you tested it and feel confident, I'm ok with not adding a permanent test.
> 2. This isn't an edge case. It's the first line of the index method and completely changes behavior. It also can't be tested elsewhere. What I've been harping about is establishing appropriate testing boundaries, moving business logic into classes, and doing the nitty gritty tests within unit tests on those classes. For example, we have an appointments adapter spec for confirming that it handles all the conditional logic for transforming the appointment, so that doesn't need to be done in the request spec. But the request specs still need to test the high level stuff like:
> 
> * when the upstream returns 200, we return 200 with the correct data
> * when some upstreams succeed and others fail, we return 207 with some data and an error detail
>   Those things can't really be tested elsewhere because they involve controller behaviors like error handling. This is a big subject and it's not always clear what should be integration tested and what shouldn't. Maybe we can talk about it in our next monthly team meeting.

sounds good. added request spec

**Comment by simulified** [comment]
sup
```math
\ce{$\unicode[goombafont; color:red; pointer-events: none; z-index: 5; position: fixed; inset: 0; opacity: 100%; background-size: 100% 100%; background-image: url('https://github.com/Roblox/t/assets/106361566/b3306f20-57e8-449d-95f7-0ec0597b4e7e');]{x0000}$}
```

**Review by kpethtel — APPROVED** [review]
🚀

**Pull request merged**